From c87c93bc66d488301e3a559148538d6130f4c85c Mon Sep 17 00:00:00 2001
From: Sohil Mehta <sohil.mehta@intel.com>
Date: Thu, 8 Sep 2022 15:58:30 -0700
Subject: [PATCH 14/28] x86/uintr: Wire up the user interrupt syscalls

Wire up the user interrupt receiver and sender related syscalls for
x86_64.

To avoid changing syscall numbers frequently during internal
development, the UINTR syscalls start at 471. This leaves a temporary
gap for upstream syscall numbers that get added.

For rest of the architectures the syscalls are not implemented.

<TODO: Reserve the syscall numbers for other architectures>

<TODO: Correct the syscall numbers before this is sent upstream. >

Signed-off-by: Sohil Mehta <sohil.mehta@intel.com>
---
 arch/x86/entry/syscalls/syscall_32.tbl | 12 ++++++++++++
 arch/x86/entry/syscalls/syscall_64.tbl | 10 ++++++++++
 include/linux/syscalls.h               | 13 +++++++++----
 include/uapi/asm-generic/unistd.h      | 25 ++++++++++++++++++++++++-
 kernel/sys_ni.c                        | 11 +++++++++++
 scripts/checksyscalls.sh               |  9 +++++++++
 6 files changed, 75 insertions(+), 5 deletions(-)

diff --git a/arch/x86/entry/syscalls/syscall_32.tbl b/arch/x86/entry/syscalls/syscall_32.tbl
index e979a3eac7a3..38aaae60e0a4 100644
--- a/arch/x86/entry/syscalls/syscall_32.tbl
+++ b/arch/x86/entry/syscalls/syscall_32.tbl
@@ -456,6 +456,8 @@
 448	i386	process_mrelease	sys_process_mrelease
 449	i386	futex_waitv		sys_futex_waitv
 450	i386	set_mempolicy_home_node		sys_set_mempolicy_home_node
+<<<<<<< conflict 1 of 1
++++++++ nynontlr 0968f96c "x86/uintr: Introduce uintr_wait() syscall" (rebased revision)
 451	i386	cachestat		sys_cachestat
 452	i386	fchmodat2		sys_fchmodat2
 453	i386	map_shadow_stack	sys_map_shadow_stack
@@ -476,3 +478,13 @@
 468	i386	file_getattr		sys_file_getattr
 469	i386	file_setattr		sys_file_setattr
 470	i386	listns			sys_listns
+# Start UINTR syscalls at 471 to avoid upstream merge conflicts during development
+471	i386	uintr_register_handler	sys_uintr_register_handler
+472	i386	uintr_unregister_handler sys_uintr_unregister_handler
+473	i386	uintr_vector_fd		sys_uintr_vector_fd
+474	i386	uintr_register_sender	sys_uintr_register_sender
+475	i386	uintr_unregister_sender	sys_uintr_unregister_sender
+476	i386	uintr_wait		sys_uintr_wait
+477	i386	uintr_register_self	sys_uintr_register_self
+488	i386	uintr_alt_stack		sys_uintr_alt_stack
+489	i386	uintr_ipi_fd		sys_uintr_ipi_fd
diff --git a/arch/x86/entry/syscalls/syscall_64.tbl b/arch/x86/entry/syscalls/syscall_64.tbl
index 8a4ac4841be6..a4aee92f07ca 100644
--- a/arch/x86/entry/syscalls/syscall_64.tbl
+++ b/arch/x86/entry/syscalls/syscall_64.tbl
@@ -395,6 +395,16 @@
 468	common	file_getattr		sys_file_getattr
 469	common	file_setattr		sys_file_setattr
 470	common	listns			sys_listns
+# Start UINTR syscalls at 471 to avoid upstream merge conflicts during development
+471	common	uintr_register_handler	sys_uintr_register_handler
+472	common	uintr_unregister_handler sys_uintr_unregister_handler
+473	common	uintr_vector_fd		sys_uintr_vector_fd
+474	common	uintr_register_sender	sys_uintr_register_sender
+475	common	uintr_unregister_sender	sys_uintr_unregister_sender
+476	common	uintr_wait		sys_uintr_wait
+477	common	uintr_register_self	sys_uintr_register_self
+478	common	uintr_alt_stack		sys_uintr_alt_stack
+479	common	uintr_ipi_fd		sys_uintr_ipi_fd
 
 #
 # Due to a historical design error, certain syscalls are numbered differently
diff --git a/include/linux/syscalls.h b/include/linux/syscalls.h
index cf84d98964b2..1d0bcf63d0a4 100644
--- a/include/linux/syscalls.h
+++ b/include/linux/syscalls.h
@@ -1006,10 +1006,15 @@ asmlinkage long sys_lsm_list_modules(u64 __user *ids, u32 __user *size, u32 flag
 
 /* x86 */
 asmlinkage long sys_ioperm(unsigned long from, unsigned long num, int on);
-
-asmlinkage long sys_uretprobe(void);
-
-asmlinkage long sys_uprobe(void);
+asmlinkage long sys_uintr_register_handler(u64 __user *handler, unsigned int flags);
+asmlinkage long sys_uintr_unregister_handler(unsigned int flags);
+asmlinkage long sys_uintr_vector_fd(u64 vector, unsigned int flags);
+asmlinkage long sys_uintr_register_sender(int uvec_fd, unsigned int flags);
+asmlinkage long sys_uintr_unregister_sender(int uvec_fd, unsigned int flags);
+asmlinkage long sys_uintr_wait(u64 usec, unsigned int flags);
+asmlinkage long sys_uintr_register_self(u64 vector, unsigned int flags);
+asmlinkage long sys_uintr_alt_stack(void __user *sp, size_t size, unsigned int flags);
+asmlinkage long sys_uintr_ipi_fd(unsigned int flags);
 
 /* pciconfig: alpha, arm, arm64, ia64, sparc */
 asmlinkage long sys_pciconfig_read(unsigned long bus, unsigned long dfn,
diff --git a/include/uapi/asm-generic/unistd.h b/include/uapi/asm-generic/unistd.h
index 942370b3f5d2..f0b9cccf3dc2 100644
--- a/include/uapi/asm-generic/unistd.h
+++ b/include/uapi/asm-generic/unistd.h
@@ -860,8 +860,31 @@ __SYSCALL(__NR_file_setattr, sys_file_setattr)
 #define __NR_listns 470
 __SYSCALL(__NR_listns, sys_listns)
 
+/*
+ * Skip syscall numbers until 470. Start UINTR syscall numbers at 471 to avoid
+ * upstream merge conflicts during internal development.
+ */
+#define __NR_uintr_register_handler 471
+__SYSCALL(__NR_uintr_register_handler, sys_uintr_register_handler)
+#define __NR_uintr_unregister_handler 472
+__SYSCALL(__NR_uintr_unregister_handler, sys_uintr_unregister_handler)
+#define __NR_uintr_vector_fd 473
+__SYSCALL(__NR_uintr_vector_fd, sys_uintr_vector_fd)
+#define __NR_uintr_register_sender 474
+__SYSCALL(__NR_uintr_register_sender, sys_uintr_register_sender)
+#define __NR_uintr_unregister_sender 475
+__SYSCALL(__NR_uintr_unregister_sender, sys_uintr_unregister_sender)
+#define __NR_uintr_wait 476
+__SYSCALL(__NR_uintr_wait, sys_uintr_wait)
+#define __NR_uintr_register_self 477
+__SYSCALL(__NR_uintr_register_self, sys_uintr_register_self)
+#define __NR_uintr_alt_stack 478
+__SYSCALL(__NR_uintr_alt_stack, sys_uintr_alt_stack)
+#define __NR_uintr_ipi_fd 479
+__SYSCALL(__NR_uintr_ipi_fd, sys_uintr_ipi_fd)
+
 #undef __NR_syscalls
-#define __NR_syscalls 471
+#define __NR_syscalls 480
 
 /*
  * 32 bit systems traditionally used different
diff --git a/kernel/sys_ni.c b/kernel/sys_ni.c
index bf5d05c635ff..180493596542 100644
--- a/kernel/sys_ni.c
+++ b/kernel/sys_ni.c
@@ -270,6 +270,17 @@ COND_SYSCALL(pkey_free);
 /* memfd_secret */
 COND_SYSCALL(memfd_secret);
 
+/* user interrupts */
+COND_SYSCALL(uintr_register_handler);
+COND_SYSCALL(uintr_unregister_handler);
+COND_SYSCALL(uintr_vector_fd);
+COND_SYSCALL(uintr_register_sender);
+COND_SYSCALL(uintr_unregister_sender);
+COND_SYSCALL(uintr_wait);
+COND_SYSCALL(uintr_register_self);
+COND_SYSCALL(uintr_alt_stack);
+COND_SYSCALL(uintr_ipi_fd);
+
 /*
  * Architecture specific weak syscall entries.
  */
diff --git a/scripts/checksyscalls.sh b/scripts/checksyscalls.sh
index 1e5d2eeb726d..38ca667b9e15 100755
--- a/scripts/checksyscalls.sh
+++ b/scripts/checksyscalls.sh
@@ -201,6 +201,15 @@ cat << EOF
 #define __IGNORE__sysctl
 #define __IGNORE_arch_prctl
 #define __IGNORE_nfsservctl
+#define __IGNORE_uintr_register_handler
+#define __IGNORE_uintr_unregister_handler
+#define __IGNORE_uintr_vector_fd
+#define __IGNORE_uintr_register_sender
+#define __IGNORE_uintr_unregister_sender
+#define __IGNORE_uintr_wait
+#define __IGNORE_uintr_register_self
+#define __IGNORE_uintr_alt_stack
+#define __IGNORE_uintr_ipi_fd
 
 /* ... including the "new" 32-bit uid syscalls */
 #define __IGNORE_lchown32
-- 
2.52.0

